src/test-com.c

```diff
@@ -1,196 +1,320 @@
+	tw = 16;
+	printf("Enter sampling time (ms) > ");
+	fflush(stdout);
+	//scanf("%d", &tw);
+	printf("Sampling time = %d ms\n", tw);
+
+	printf("Enter File name > ");
+	fflush(stdout);
+	//scanf("%s", fname);
+	//fd = fopen(fname, "w");
+	//fd = fopen("test.txt", "w");
+	//if (!fd)
+	//	goto over;
+	#if defined(__QNX__)
+        slogf(0, _SLOG_INFO, "Started  input: %s fd = %d\n", devname_in, fd);
+	    slogf(0, _SLOG_INFO, "        output: %s fd = %d\n", devname_out, fd);
+	#endif
+
+	// Set Commport Settings
+	SetComAttr(fd);
+
+	// Read Data
+	printf("=== record data ===\n");
+	clk0 = clock() / (CLOCKS_PER_SEC / 1000);
+	clkb = 0;
+	clkb2 = 0;
+	num = 0;
+
+	// Data Requests
+	struct timeval wr0, wr1, loop0, loop1, fscan0, fscan1;
+	gettimeofday(&wr0, NULL);
+	n = write(fd, "R", 1);
+	gettimeofday(&wr1, NULL);
+	// tcdrain(fd); // Old legacy that is no longer needed. Besideds that, on QNX6.5.0 this is found taking unnecessarily too long (say 100msec).
+#define DELTA_SEC(start, end) (end.tv_sec - start.tv_sec + (end.tv_usec - start.tv_usec)/1e6)
+	printf("write took %7.3f [msec]\n", DELTA_SEC(wr0, wr1) * 1000);
+	printf("write data (ret %d)\n", n);
+
+	while (true) {
+		// Wait for Sampling rate
+		while (true) {
+			clk = clock() / (CLOCKS_PER_SEC / 1000) - clk0;
+			if (clk >= clkb + tw) {
+				clkb = clk / tw * tw;
+				break;
+			}
+		}
+
+		// Data Request
+		gettimeofday(&wr0, NULL);
+		n = write(fd, "R", 1);
+		gettimeofday(&wr1, NULL);
+		// tcdrain(fd);  //  The tcdrain() function waits until all output has been physically transmitted to the device associated with fd, or until a signal is received.
+		printf("write took %7.3f [msec]\n", DELTA_SEC(wr0, wr1) * 1000);
+		printf("write data (ret %d)\n", n);
+
+		// Get Single data
+#define DATA_LENGTH 27
+		len = 0;
+		bzero(str, 27);  // Setting the first 27 bytes of the area starting `str` to zero.
+		while (len < DATA_LENGTH) {
+			gettimeofday(&loop0, NULL);
+			n = read(fd, str + len, DATA_LENGTH - len);
+			printf("read data (ret %d, %d bytes read))\n", n, len + n);
+			if (n > 0) {
+				len += n;
+			} else {
+				char *pmesg = strerror(errno);
+				fprintf(stderr, "failed to read data (%d): %s (%d)\n", n, pmesg,
+						errno);
+				usleep(10000);
+				//goto loop_exit;
+			}
+			gettimeofday(&loop1, NULL);
+			printf("Loop (read) took %7.3f [msec]\n", DELTA_SEC(loop0, loop1) * 1000);
+		}
+		//goto skip;
+		loop_exit: {
+			int i;
+			for (i = 0; i < DATA_LENGTH; i++) {
+				fprintf(stderr, "%02x:", 0x0000ff & str[i]);
+			}
+			printf("\n");
+		}
+
+		gettimeofday(&fscan0, NULL);
+		sscanf(str, "%1d%4hx%4hx%4hx%4hx%4hx%4hx", &tick, &data[0], &data[1],
+				&data[2], &data[3], &data[4], &data[5]);
+
+		sprintf(str, "%05d,%d,%05d,%05d,%05d,%05d,%05d,%05d\n", clk / tw * tw,
+				tick, data[0], data[1], data[2], data[3], data[4], data[5]);
+		gettimeofday(&fscan1, NULL);
+		printf("sscanf, sprintf took %7.3f [msec]\n", DELTA_SEC(fscan0, fscan1) * 1000);
+
+		fprintf(stderr, str);
```

Comment:
Build is failing on ROS buildfarm ([e.g.](http://build.ros.org/view/Kbin_uX64/job/Kbin_uX64__dynpick_driver__ubuntu_xenial_amd64__binary/1/console)):

```
15:08:55 [ 80%] Building C object CMakeFiles/dynpick_com_tester.dir/src/test-com.c.o
15:08:55 /usr/lib/ccache/x86_64-linux-gnu-gcc  -DROSCONSOLE_BACKEND_LOG4CXX -DROS_PACKAGE_NAME=\"dynpick_driver\" -I/opt/ros/kinetic/include  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2    -o CMakeFiles/dynpick_com_tester.dir/src/test-com.c.o   -c /tmp/binarydeb/ros-kinetic-dynpick-driver-0.0.10/src/test-com.c
15:08:55 /tmp/binarydeb/ros-kinetic-dynpick-driver-0.0.10/src/test-com.c: In function ‘main’:
15:08:55 /tmp/binarydeb/ros-kinetic-dynpick-driver-0.0.10/src/test-com.c:126:2: warning: implicit declaration of function ‘SetComAttr’ [-Wimplicit-function-declaration]
15:08:55   SetComAttr(fd);
15:08:55   ^
15:08:55 /tmp/binarydeb/ros-kinetic-dynpick-driver-0.0.10/src/test-com.c:201:18: error: format not a string literal and no format arguments [-Werror=format-security]
15:08:55    fprintf(stderr, str);
15:08:55                   ^
15:08:55 /tmp/binarydeb/ros-kinetic-dynpick-driver-0.0.10/src/test-com.c:210:11: error: format not a string literal and no format arguments [-Werror=format-security]
15:08:55     printf(str);
15:08:55            ^
15:08:55 /tmp/binarydeb/ros-kinetic-dynpick-driver-0.0.10/src/test-com.c:226:2: warning: implicit declaration of function ‘exit’ [-Wimplicit-function-declaration]
15:08:55   exit(0);   //scanf("%s", str);
15:08:55   ^
15:08:55 /tmp/binarydeb/ros-kinetic-dynpick-driver-0.0.10/src/test-com.c:226:2: warning: incompatible implicit declaration of built-in function ‘exit’
15:08:55 /tmp/binarydeb/ros-kinetic-dynpick-driver-0.0.10/src/test-com.c:226:2: note: include ‘<stdlib.h>’ or provide a declaration of ‘exit’
15:08:55 /tmp/binarydeb/ros-kinetic-dynpick-driver-0.0.10/src/test-com.c:93:2: warning: ignoring return value of ‘scanf’, declared with attribute warn_unused_result [-Wunused-result]
15:08:55   scanf("%d", &comNo);
15:08:55   ^
15:08:55 cc1: some warnings being treated as errors
15:08:55 make[4]: *** [CMakeFiles/dynpick_com_tester.dir/src/test-com.c.o] Error 1
15:08:55 CMakeFiles/dynpick_com_tester.dir/build.make:65: recipe for target 'CMakeFiles/dynpick_com_tester.dir/src/test-com.c.o' failed
15:08:55 make[3]: *** [CMakeFiles/dynpick_com_tester.dir/all] Error 2
```